src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php

```diff
@@ -79,14 +82,38 @@ public function decode(array $encodedEnvelope): Envelope
         }
 
         try {
-            $message = $this->serializer->deserialize($encodedEnvelope['body'], $encodedEnvelope['headers']['type'], $this->format, $context);
+            $message = $this->deserialize($encodedEnvelope, $context);
         } catch (ExceptionInterface $e) {
             throw new MessageDecodingFailedException('Could not decode message: '.$e->getMessage(), $e->getCode(), $e);
         }
 
         return new Envelope($message, $stamps);
     }
 
+    private function deserialize(array $encodedEnvelope, array $context)
+    {
+        $resolver = $context[self::TYPE_RESOLVER] ?? new HeaderTypeResolver();
+
+        if ($resolver instanceof TypeResolverInterface) {
+            $type = $resolver->resolve($encodedEnvelope);
+
+            return $this->serializer->deserialize($encodedEnvelope['body'], $type, $this->format, $context);
+        }
+
+        if (!$resolver instanceof DecodedAwareTypeResolverInterface) {
+            throw new MessageDecodingFailedException(sprintf(sprintf('"%s" must be an instance of "%s" or "%s", "%s" given.', self::TYPE_RESOLVER, TypeResolverInterface::class, DecodedAwareTypeResolverInterface::class, \is_object($resolver) ? \get_class($resolver) : \gettype($resolver))));
```

Comment:
It looks like you have two `sprintf` nested here. That should be fixed.

Reply:
Indeed :-s. Thank you 🙂